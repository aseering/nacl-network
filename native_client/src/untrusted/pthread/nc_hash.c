/*
 * Copyright 2008, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


/*
 * A hash table for pthreads library
 */

#include <unistd.h>
#include "nc_hash.h"
#include "pthread.h"

static uint32_t hash_get_bucket_id(struct NaClHashTable *table,
                                   uint32_t id) {
  return id % table->number_of_buckets;
}

int __nacl_hash_init(struct NaClHashTable *table) {
  uint32_t i = 0;

  table->number_of_buckets = sizeof(table->buckets) / sizeof(table->buckets[0]);
  for (i = 0; i < table->number_of_buckets; ++i) {
    LIST_INIT(&table->buckets[i]);
  }
  return 1;
}

void __nacl_hash_destroy(struct NaClHashTable *table) {
  /*
   * Nothing to free - everything is static
   * TODO(gregoryd): what do we do when there
   * are still some entries in the table?
   */
}

int __nacl_hash_insert_mu(struct NaClHashTable *table,
                          nc_hash_entry_t *entry) {
  uint32_t bucket_id = hash_get_bucket_id(table, entry->id_function(entry));
  LIST_INSERT_HEAD(&table->buckets[bucket_id], entry, list_entries);
  return 0;
}

nc_hash_entry_t *__nacl_hash_find_id_mu(struct NaClHashTable *table,
                                        uint32_t id) {
  uint32_t bucket_id = hash_get_bucket_id(table, id);
  nc_hash_entry_t *iter;
  LIST_FOREACH(iter, &table->buckets[bucket_id], list_entries) {
    if (iter->id_function(iter) == id) {
      return iter;
    }
  }
  return NULL;
}

int __nacl_hash_id_exists_mu(struct NaClHashTable *table,
                             uint32_t id) {
  if (NULL == __nacl_hash_find_id_mu(table, id)) {
    return 0;
  }
  return 1;
}

nc_hash_entry_t *__nacl_hash_remove_mu(struct NaClHashTable  *table,
                                       uint32_t id) {
  uint32_t bucket_id = hash_get_bucket_id(table, id);
  nc_hash_entry_t *iter;
  LIST_FOREACH(iter, &table->buckets[bucket_id], list_entries) {
    if (iter->id_function(iter) == id) {
      LIST_REMOVE(iter, list_entries);
      return iter;
    }
  }
  return NULL;
}
