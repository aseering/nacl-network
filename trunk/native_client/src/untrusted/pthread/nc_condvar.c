/*
 * Copyright 2008, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


/*
 * Native Client condition variable API
 */

#include <unistd.h>

#include "native_client/src/untrusted/nacl/syscall_bindings_trampoline.h"

#include "native_client/src/untrusted/pthread/pthread.h"
#include "native_client/src/untrusted/pthread/pthread_types.h"


/* TODO(gregoryd): make this static?  */
void pthread_cond_validate(pthread_cond_t* cond) {
  nc_spinlock_lock(&cond->spinlock);

  if (NC_INVALID_HANDLE == cond->handle) {
    pthread_cond_init(cond, NULL);
  }

  nc_spinlock_unlock(&cond->spinlock);
}


/*
 * Initialize condition variable COND using attributes ATTR, or use
 * the default values if later is NULL.
 */
int pthread_cond_init (pthread_cond_t *cond,
                       pthread_condattr_t *cond_attr) {
  cond->handle = NACL_SYSCALL(cond_create)();

  /* 0 for success, 1 for failure */
  return (cond->handle < 0);
}

/*
 * Destroy condition variable COND.
 */
int pthread_cond_destroy (pthread_cond_t *cond) {
  int retval;
  pthread_cond_validate(cond);
  retval = close(cond->handle);
  cond->handle = NC_INVALID_HANDLE;
  return retval;
}

/*
 * Wake up one thread waiting for condition variable COND.
 */
int pthread_cond_signal (pthread_cond_t *cond) {
  pthread_cond_validate(cond);
  return NACL_SYSCALL(cond_signal)(cond->handle);
}


int pthread_cond_broadcast (pthread_cond_t *cond) {
  pthread_cond_validate(cond);
  return NACL_SYSCALL(cond_broadcast)(cond->handle);
}

int pthread_cond_wait (pthread_cond_t *cond,
                       pthread_mutex_t *mutex) {
  pthread_cond_validate(cond);
  return NACL_SYSCALL(cond_wait)(cond->handle, mutex->mutex_handle);
}

int pthread_cond_timedwait_abs(pthread_cond_t *cond,
                               pthread_mutex_t *mutex,
                               struct timespec *abstime) {
  pthread_cond_validate(cond);
  return NACL_SYSCALL(cond_timed_wait_abs)(cond->handle,
                                           mutex->mutex_handle,
                                           abstime);
}

int nc_pthread_condvar_ctor(pthread_cond_t *cond) {
  cond->handle = NC_INVALID_HANDLE;
  cond->spinlock = 0;
  return 1;
}
