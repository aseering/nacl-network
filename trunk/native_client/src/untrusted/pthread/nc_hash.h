/*
 * Copyright 2008, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


/*
 * A hash table for phtreads library
 */

#ifndef NATIVE_CLIENT_SRC_UNTRUSTED_PTHREAD_NC_HASH_H_
#define NATIVE_CLIENT_SRC_UNTRUSTED_PTHREAD_NC_HASH_H_ 1

#include <stdint.h>
#include "pthread.h"

/*
 * Note that this file is implementing templates by macros, as C doesn't
 * support templates.
 */
#define HASH_BUCKETS_NUMBER 50

typedef struct nc_hash_entry {
  LIST_ENTRY(nc_hash_entry) list_entries;
  /* a basic vtbl */
  uint32_t (*id_function)(struct nc_hash_entry *);
} nc_hash_entry_t;

LIST_HEAD(HashListHead, nc_hash_entry);

struct NaClHashTable {
  uint32_t number_of_buckets;
  struct HashListHead buckets[HASH_BUCKETS_NUMBER];
};

int __nacl_hash_init(struct NaClHashTable *table);

void __nacl_hash_destroy(struct NaClHashTable *table);

int __nacl_hash_insert_mu(struct NaClHashTable *table,
                          nc_hash_entry_t *entry);

nc_hash_entry_t *__nacl_hash_find_id_mu(struct NaClHashTable *table,
                                        uint32_t id);

int __nacl_hash_id_exists_mu(struct NaClHashTable *table,
                             uint32_t id);

nc_hash_entry_t *__nacl_hash_remove_mu(struct NaClHashTable  *table,
                                       uint32_t id);

#define HASH_INIT(table)  \
  __nacl_hash_init(table)

#define HASH_TABLE_INSERT(table, entry, hash_entry_name) \
  __nacl_hash_insert_mu(table, &entry->hash_entry_name)

#define HASH_ENTRY_TO_ENTRY_TYPE(hash_entry, entry_type, hash_entry_name) \
  ((hash_entry)?((struct entry_type *) \
  ((char*)hash_entry - offsetof(struct entry_type, hash_entry_name))): NULL)

#define HASH_FIND_ID(table, id, entry_type, hash_entry_name) \
  HASH_ENTRY_TO_ENTRY_TYPE(__nacl_hash_find_id_mu(table, id), \
                           entry_type, hash_entry_name)

#define HASH_ID_EXISTS(table, id) \
  __nacl_hash_id_exists_mu(table, id)

#define HASH_REMOVE(table, id, entry_type, hash_entry_name) \
  HASH_ENTRY_TO_ENTRY_TYPE(__nacl_hash_remove_mu(table, id), \
                           entry_type, hash_entry_name)

#endif  /* NATIVE_CLIENT_SRC_UNTRUSTED_PTHREAD_NC_HASH_H_ */
