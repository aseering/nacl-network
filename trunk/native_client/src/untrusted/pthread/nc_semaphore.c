/*
 * Copyright 2008, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


/*
* Native Client semaphore API
*/

#include <unistd.h>
#include <errno.h>

#include "native_client/src/untrusted/nacl/syscall_bindings_trampoline.h"

#include "native_client/src/untrusted/pthread/pthread.h"
#include "native_client/src/untrusted/pthread/pthread_types.h"
#include "native_client/src/untrusted/pthread/semaphore.h"

/* Initialize semaphore  */
int sem_init (sem_t *sem, int pshared, unsigned int value) {
  if (pshared) {
    /* we don's support shared semaphores yet */
    errno = EINVAL;
    return -1;
  }
  sem->handle = NACL_SYSCALL(sem_create)(value);
  return (sem->handle < 0);
}

int sem_destroy (sem_t *sem) {
  int retval = close(sem->handle);
  sem->handle = NC_INVALID_HANDLE;
  return retval;
}

int sem_wait (sem_t *sem) {
  return NACL_SYSCALL(sem_wait)(sem->handle);
}

int sem_post (sem_t *sem) {
  int32_t rv = NACL_SYSCALL(sem_post)(sem->handle);
  if (0 != rv) {
    errno = rv;
    return -1;
  }
  return rv;
}
