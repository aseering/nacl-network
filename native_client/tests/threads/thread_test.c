/*
 * Copyright 2008, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Testing suite for NativeClient threads
 */

#include <pthread.h>
#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * We cannot have more than just shy of 8192 threads running
 * simultaneously.  This is a NaCl architectural limitation.
 */

const int kTestLoops = 10000;
/* Macros so we can use it for array dimensions in ISO C90 */
#define NUM_THREADS 5

__thread int tls_var = 5;
int ready = 0;

int errors = 0;

#define PRINT_ERROR { printf("Error on line %d\n", __LINE__); errors++; }

#define EXPECT_EQ(A, B) if ((A)!=(B)) PRINT_ERROR
#define EXPECT_NE(A, B) if ((A)==(B)) PRINT_ERROR
#define EXPECT_GE(A, B) if ((A)<(B)) PRINT_ERROR
#define EXPECT_LE(A, B) if ((A)>(B)) PRINT_ERROR

#define PRINT_TEST_OK printf("%s - OK\n", __FUNCTION__)
#define PRINT_TEST_FAILED printf("%s - FAILED\n", __FUNCTION__)

#define TEST_FUNCTION_START int local_error = errors; \
                            printf("%s - running\n", __FUNCTION__);

#define TEST_FUNCTION_END if (local_error == errors) \
                            PRINT_TEST_OK; \
                          else \
                            PRINT_TEST_FAILED;

struct SYNC_DATA {
  pthread_mutex_t mutex;
  pthread_cond_t cv;
};

typedef void* (*ThreadFunction)(void *state);

void* FastThread(void *userdata) {
  /* do nothing and immediately exit */
  return 0;
}

/* creates and waits via pthread_join() for thread to exit */
void CreateWithJoin(ThreadFunction func, void *state) {
  pthread_t thread_id;
  void *thread_ret;
  int p = pthread_create(&thread_id, NULL, func, state);
  EXPECT_EQ(0, p);

  /* wait for thread to exit */
  p = pthread_join(thread_id, &thread_ret);
  EXPECT_EQ(0, p);
}


/* creates, but does not wait for thread to exit */
void CreateWithoutJoin() {
  pthread_t thread_id;
  int p = pthread_create(&thread_id, NULL, FastThread, NULL);
  EXPECT_EQ(0, p);
  /* intentionally no pthread_join() */
}


/* creates as detached thread, cannot join */
void CreateDetached() {
  int p;
  pthread_t thread_id;
  pthread_attr_t attr;
  pthread_attr_init(&attr);
  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
  p = pthread_create(&thread_id, &attr, FastThread, NULL);
  EXPECT_EQ(0, p);
  /* cannot join on detached thread */
}

void* TlsThread(void* state) {
  struct SYNC_DATA *sync_data = (struct SYNC_DATA *)state;
  pthread_mutex_lock(&sync_data->mutex);
  tls_var = 8;
  ready = 1;
  pthread_cond_signal(&sync_data->cv);
  pthread_mutex_unlock(&sync_data->mutex);
  return (void*)33;
}

void TestTlsAndSync() {
  int i = 5;
  void* thread_rv;
  pthread_t thread_id;
  pthread_attr_t attr;
  struct SYNC_DATA sync_data;

  TEST_FUNCTION_START;

  pthread_mutex_init(&sync_data.mutex, NULL);
  pthread_cond_init(&sync_data.cv, NULL);
  pthread_attr_init(&attr);
  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
  i = pthread_create(&thread_id, &attr, TlsThread, &sync_data);
  EXPECT_EQ(0, i);

  EXPECT_EQ(5, tls_var);
  pthread_mutex_lock(&sync_data.mutex);

  while (!ready) {
    pthread_cond_wait(&sync_data.cv, &sync_data.mutex);
  }
  i = pthread_join(thread_id, &thread_rv);
  EXPECT_NE(0, i);  /* join should fail since the thread was created detached */
  EXPECT_EQ(5, tls_var);

  pthread_mutex_unlock(&sync_data.mutex);
  EXPECT_EQ(5, tls_var);
  TEST_FUNCTION_END;
}

void TestManyThreadsJoinable() {
  int i;
  TEST_FUNCTION_START;
  for (i = 0; i < kTestLoops; i++) {
    CreateWithJoin(FastThread, NULL);
  }
  TEST_FUNCTION_END;
}



/* this test actually consumes all thread ids - no other threads will succeed */
#if 0
  for (i = 0; i < NC_MAX_THREADS-1; i++) {
    /* without join we are limited by the max number of threads */
    CreateWithoutJoin();
  }
#endif

void TestManyThreadsDetached() {
  int i;
  TEST_FUNCTION_START;
  for (i = 0; i < kTestLoops; i++) {
    CreateDetached();
  }
  TEST_FUNCTION_END;
}

void* SemaphoresThread(void *state) {
  sem_t* sem = (sem_t*) state;
  int i = 0, rv;
  for (i = 0; i < kTestLoops; i++) {
    rv = sem_wait(&sem[0]);
    EXPECT_EQ(0, rv);
    rv = sem_post(&sem[1]);
    EXPECT_EQ(0, rv);
  }
  EXPECT_EQ(kTestLoops, i);
  return 0;
}

void TestSemaphores() {
  int i = 5;
  int rv;
  pthread_t thread_id;
  pthread_attr_t attr;
  sem_t sem[2];
  TEST_FUNCTION_START;
  sem_init(&sem[0], 0, 0);
  sem_init(&sem[1], 0, 0);

  pthread_attr_init(&attr);
  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);

  i = pthread_create(&thread_id, &attr, SemaphoresThread, sem);
  EXPECT_EQ(0, i);

  for (i = 0; i < kTestLoops; i++) {
    rv = sem_post(&sem[0]);
    EXPECT_EQ(0, rv);
    rv = sem_wait(&sem[1]);
    EXPECT_EQ(0, rv);
  }
  sem_destroy(&sem[0]);
  sem_destroy(&sem[1]);
  TEST_FUNCTION_END;
}

pthread_once_t once_control = PTHREAD_ONCE_INIT;

void pthread_once_routine() {
  static AtomicWord count = 0;
  AtomicWord res;
  res = AtomicIncrement(&count, 1);
  EXPECT_LE(res, 1);
}

void* OnceThread(void *userdata) {
  pthread_once(&once_control, pthread_once_routine);
  return 0;
}


void TestPthreadOnce() {
  int i;
  TEST_FUNCTION_START;
  for (i = 0; i < kTestLoops; i++) {
    pthread_t thread_id;
    pthread_attr_t attr;
    int p;
    pthread_attr_init(&attr);
    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
    p = pthread_create(&thread_id, &attr, OnceThread, NULL);
    EXPECT_EQ(0, p);
  }
  TEST_FUNCTION_END;
}

void* RecursiveLockThread(void *state) {
  int rv;
  int i;
  pthread_mutex_t *lock = state;

  for (i = 0; i < kTestLoops; ++i) {
    rv = pthread_mutex_lock(lock);
    EXPECT_EQ(0, rv);
  }

  for (i = 0; i < kTestLoops; ++i) {
    rv = pthread_mutex_unlock(lock);
    EXPECT_EQ(0, rv);
  }
  return 0;
}

void TestRecursiveMutex() {
  pthread_mutexattr_t attr;
  pthread_mutex_t mutex = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
  int rv;
  pthread_t tid[NUM_THREADS];
  int i = 0;
  TEST_FUNCTION_START;

  for (i = 0; i < NUM_THREADS; ++i) {
    int rv = pthread_create(&tid[i], NULL, RecursiveLockThread, &mutex);
    EXPECT_EQ(0, rv);
  }

  for (i = 0; i < NUM_THREADS; ++i) {
    pthread_join(tid[i], NULL);
  }

  rv = pthread_mutex_lock(&mutex);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_trylock(&mutex);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_unlock(&mutex);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_unlock(&mutex);
  EXPECT_EQ(0, rv);

  rv = pthread_mutex_destroy(&mutex);
  EXPECT_EQ(0, rv);
  memset(&mutex, 0, sizeof(mutex));

  rv = pthread_mutexattr_init(&attr);
  EXPECT_EQ(0, rv);
  rv = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE_NP);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_init(&mutex, &attr);
  EXPECT_EQ(0, rv);

  for (i = 0; i < NUM_THREADS; ++i) {
    int rv = pthread_create(&tid[i], NULL, RecursiveLockThread, &mutex);
    EXPECT_EQ(0, rv);
  }

  for (i = 0; i < NUM_THREADS; ++i) {
    pthread_join(tid[i], NULL);
  }

  TEST_FUNCTION_END;
}

void TestErrorCheckingMutex() {
  pthread_mutexattr_t attr;
  pthread_mutex_t mutex;
  int rv;
  TEST_FUNCTION_START;
  rv = pthread_mutexattr_init(&attr);
  EXPECT_EQ(0, rv);
  rv = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK_NP);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_init(&mutex, &attr);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_unlock(&mutex);
  EXPECT_NE(0, rv);
  rv = pthread_mutex_lock(&mutex);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_trylock(&mutex);
  EXPECT_NE(0, rv);
  rv = pthread_mutex_unlock(&mutex);
  EXPECT_EQ(0, rv);
  rv = pthread_mutex_unlock(&mutex);
  EXPECT_NE(0, rv);
  TEST_FUNCTION_END;
}

void tsd_destructor(void *arg) {
  *(int*)arg += 1;
}

pthread_key_t tsd_key;

void* TsdThread(void *state) {
  int rv;
  rv = pthread_setspecific(tsd_key, state);
  EXPECT_EQ(0, rv);
  pthread_exit((void*) 5);
  return 0;
}

void TestTSD() {
  int rv;
  void* ptr;
  int destructor_count = 0;
  TEST_FUNCTION_START;
  rv = pthread_key_create(&tsd_key, tsd_destructor);
  EXPECT_EQ(0, rv);

  rv = pthread_setspecific(tsd_key, &rv);
  EXPECT_EQ(0, rv);

  ptr = pthread_getspecific(tsd_key);
  EXPECT_EQ(ptr, &rv);

  CreateWithJoin(TsdThread, &destructor_count);
  EXPECT_EQ(1, destructor_count);

  rv = pthread_key_delete(tsd_key);
  EXPECT_EQ(0, rv);

  TEST_FUNCTION_END;
}

void* MallocThread(void *userdata) {
  void* ptr;
  int i;
  for (i = 0; i < kTestLoops; ++i) {
    ptr = (void*) malloc(16);
    EXPECT_NE(NULL, ptr);
  }
  return ptr;
}

void TestMalloc() {
  int i = 0;
  pthread_t tid[NUM_THREADS];
  TEST_FUNCTION_START;
  for (i = 0; i < NUM_THREADS; ++i) {
    int rv = pthread_create(&tid[i], NULL, MallocThread, NULL);
    EXPECT_EQ(0, rv);
  }

  for (i = 0; i < NUM_THREADS; ++i) {
    pthread_join(tid[i], NULL);
  }

  TEST_FUNCTION_END;
}

void* ReallocThread(void *userdata) {
  void* ptr;
  int i;
  ptr = (void*) malloc(16);
  for (i = 0; i < kTestLoops; ++i) {
    ptr = (void*)realloc(ptr, 32);
    EXPECT_NE(NULL, ptr);
    ptr = (void*)realloc(ptr, 64000);
    EXPECT_NE(NULL, ptr);
    ptr = (void*)realloc(ptr, 64);
    EXPECT_NE(NULL, ptr);
    ptr = (void*)realloc(ptr, 32000);
    EXPECT_NE(NULL, ptr);
    ptr = (void*)realloc(ptr, 256);
    EXPECT_NE(NULL, ptr);
  }

  return ptr;
}

void TestRealloc() {
  pthread_t tid[NUM_THREADS];
  int i = 0;
  TEST_FUNCTION_START;
  for (i = 0; i < NUM_THREADS; ++i) {
    int rv = pthread_create(&tid[i], NULL, ReallocThread, NULL);
    EXPECT_EQ(0, rv);
  }

  for (i = 0; i < NUM_THREADS; ++i) {
    pthread_join(tid[i], NULL);
  }

  TEST_FUNCTION_END;
}

int main(int argc, char **argv) {
  TestTlsAndSync();
  TestManyThreadsJoinable();
  TestManyThreadsDetached();
  TestSemaphores();
  TestPthreadOnce();
  TestRecursiveMutex();
  TestErrorCheckingMutex();
  TestTSD();
  TestMalloc();
  TestRealloc();
  return errors;
}
