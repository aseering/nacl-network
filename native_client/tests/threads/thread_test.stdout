TestTlsAndSync - running
TestTlsAndSync - OK
TestManyThreadsJoinable - running
TestManyThreadsJoinable - OK
TestManyThreadsDetached - running
TestManyThreadsDetached - OK
TestSemaphores - running
TestSemaphores - OK
TestPthreadOnce - running
TestPthreadOnce - OK
TestRecursiveMutex - running
TestRecursiveMutex - OK
TestErrorCheckingMutex - running
TestErrorCheckingMutex - OK
TestTSD - running
TestTSD - OK
TestMalloc - running
TestMalloc - OK
TestRealloc - running
TestRealloc - OK
